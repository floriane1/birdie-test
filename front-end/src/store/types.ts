export interface Event {
  id: string;
  event_type: EventType;
  care_recipient_id: string;
  timestamp: Date;
  visit_id?: string;
  caregiver_id?: string;
  observation_event_id?: string;
  // tslint:disable-next-line:no-any
  [key: string]: any;
}

export enum EventType {
  ALERT_QUALIFIED = 'alert_qualified',
  ALERT_RAISED = 'alert_raised',
  CATHETER_OBSERVATION = 'catheter_observation',
  CHECK_IN = 'check_in',
  CHECK_OUT = 'check_out',
  CONCERN_RAISED = 'concern_raised',
  FLUID_INTAKE_OBSERVATION = 'fluid_intake_observation',
  FOOD_INTAKE_OBSERVATION = 'food_intake_observation',
  GENERAL_OBSERVATION = 'general_observation',
  INCONTINENCE_PAD_OBSERVATION = 'incontinence_pad_observation',
  MEDICATION_SCHEDULE_CREATED = 'medication_schedule_created',
  MEDICATION_SCHEDULE_UPDATED = 'medication_schedule_updated',
  MENTAL_HEALTH_OBSERVATION = 'mental_health_observation',
  MOOD_OBSERVATION = 'mood_observation',
  NO_MEDICATION_OBSERVATION_RECEIVED = 'no_medication_observation_received',
  PHYSICAL_HEALTH_OBSERVATION = 'physical_health_observation',
  REGULAR_MEDICATION_MAYBE_TAKEN = 'regular_medication_maybe_taken',
  REGULAR_MEDICATION_NOT_TAKEN = 'regular_medication_not_taken',
  REGULAR_MEDICATION_PARTIALLY_TAKEN = 'regular_medication_partially_taken',
  REGULAR_MEDICATION_TAKEN = 'regular_medication_taken',
  TASK_COMPLETED = 'task_completed',
  TASK_COMPLETION_REVERTED = 'task_completion_reverted',
  TASK_SCHEDULE_CREATED = 'task_schedule_created',
  TOILET_VISIT_RECORDED = 'toilet_visit_recorded',
  VISIT_CANCELLED = 'visit_cancelled',
  VISIT_COMPLETED = 'visit_completed',
}
