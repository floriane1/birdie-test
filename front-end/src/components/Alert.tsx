import * as React from 'react';
import styled from 'styled-components';
import { Event } from '@App/store/types.ts';
import Title from './Title';
import SubTitle from './SubTitle';
import { CheckCircle, ExclamationTriangle } from '@styled-icons/fa-solid';

type AlertProps = {
  events: Event[];
  alerts: Event[];
};

const AlertContainer = styled.div`
  width: 85%;
  position: relative;
  justify-content: center;
  align-items: center;
  display: flex;
  flex-direction: row;
  margin: 30px 0px;
`;

const IconContainer = styled.div`
  margin-right: 20px;
  svg {
    width: 90px;
    height: 90px;
  }
`;

const RedExclamationTriangle = styled(ExclamationTriangle)`
  color: #ff0000;
`;
const GreenCheckCircle = styled(CheckCircle)`
  color: #39d576;
`;

const Alert = ({ events, alerts }: AlertProps) => {
  return (
    <AlertContainer>
      <IconContainer>
        {alerts.length > 0 ? <RedExclamationTriangle /> : <GreenCheckCircle />}
      </IconContainer>
      <div>
        {events.length > 0 && (
          <Title>
            {alerts.length}
            {alerts.length > 1 ? ' alerts have ' : ' alert has '}
            been raised since {events[events.length - 1].timestamp}
          </Title>
        )}
        {alerts.length > 0 && (
          <SubTitle>
            {alerts.length > 1 ? 'The last one was ' : 'It was '}
            {alerts[0].timestamp}
          </SubTitle>
        )}
      </div>
    </AlertContainer>
  );
};

export default Alert;
