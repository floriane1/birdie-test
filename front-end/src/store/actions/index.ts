import { Event } from '../types';
import {
  FetchEventsRequestAction,
  FetchEventsSuccessAction,
  FetchEventsErrorAction,
  FETCH_EVENTS_REQUEST,
  FETCH_EVENTS_SUCCESS,
  FETCH_EVENTS_ERROR,
} from '../types/index';

export function fetchEventsRequest(
  careRecipientId: string
): FetchEventsRequestAction {
  return {
    type: FETCH_EVENTS_REQUEST,
    payload: careRecipientId,
  };
}

export function fetchEventsSuccess(payload: Event[]): FetchEventsSuccessAction {
  return {
    type: FETCH_EVENTS_SUCCESS,
    payload,
  };
}

export function fetchEventsError(): FetchEventsErrorAction {
  return {
    type: FETCH_EVENTS_ERROR,
  };
}
