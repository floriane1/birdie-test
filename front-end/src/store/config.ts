export const CARE_RECIPIENT_DEFAULT_ID: string =
  'ad3512a6-91b1-4d7d-a005-6f8764dd0111';
