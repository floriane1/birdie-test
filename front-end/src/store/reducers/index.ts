import { combineReducers } from 'redux';
import { Event } from '../types';
import {
  FetchEventsAction,
  FETCH_EVENTS_REQUEST,
  FETCH_EVENTS_SUCCESS,
  FETCH_EVENTS_ERROR,
} from '../types/index';

export type RootState = Readonly<{
  events: Event[];
  loading: boolean;
  error: boolean;
}>;

export const initialState: RootState = {
  events: [],
  loading: false,
  error: false,
};

export const eventsReducer = (
  state: RootState = initialState,
  action: FetchEventsAction
): RootState => {
  switch (action.type) {
    case FETCH_EVENTS_REQUEST:
      return { ...state, loading: true, error: false };
    case FETCH_EVENTS_SUCCESS:
      return { ...state, events: action.payload, loading: false, error: false };
    case FETCH_EVENTS_ERROR:
      return { ...state, loading: false, error: true };
    default:
      return state;
  }
};

export const rootReducer = combineReducers<RootState>({ eventsReducer });
