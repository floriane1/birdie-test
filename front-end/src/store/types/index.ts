import { Event } from '../types';

export const FETCH_EVENTS_REQUEST = 'FETCH_EVENTS_REQUEST';
export const FETCH_EVENTS_SUCCESS = 'FETCH_EVENTS_SUCCESS';
export const FETCH_EVENTS_ERROR = 'FETCH_EVENTS_ERROR';

export interface FetchEventsRequestAction {
  type: typeof FETCH_EVENTS_REQUEST;
  payload: string;
}

export interface FetchEventsSuccessAction {
  type: typeof FETCH_EVENTS_SUCCESS;
  payload: Event[];
}

export interface FetchEventsErrorAction {
  type: typeof FETCH_EVENTS_ERROR;
  payload: string;
}

export type FetchEventsAction =
  | FetchEventsRequestAction
  | FetchEventsSuccessAction
  | FetchEventsErrorAction;
