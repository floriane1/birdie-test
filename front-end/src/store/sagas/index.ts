import { call, put, all, takeLatest } from 'redux-saga/effects';
import axios from 'axios';
import { FetchEventsRequestAction, FETCH_EVENTS_REQUEST } from '../types/index';
import { fetchEventsSuccess, fetchEventsError } from '../actions';
import { Event } from '../types';

export function* fetchEvents(action: FetchEventsRequestAction) {
  try {
    const query = `/care-recipients/${action.payload}/events`;
    const results = yield call(axios.get, query);
    const events: Event[] = results.data.map((event: Event) => ({
      ...event,
      timestamp: new Date(
        new Date(event.timestamp).setMonth(
          // TO-DO: change when events data more recent
          new Date(event.timestamp).getMonth() + 17
        )
      ),
    }));
    yield put(fetchEventsSuccess(events));
  } catch (error) {
    yield put(fetchEventsError(error.message));
  }
}

export function* watchFetchEvents() {
  yield takeLatest(FETCH_EVENTS_REQUEST, fetchEvents);
}

function* initSaga() {
  yield all([watchFetchEvents()]);
}

export default initSaga;
