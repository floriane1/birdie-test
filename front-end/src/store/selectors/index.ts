import { createSelector } from 'reselect';
import { RootState } from '../reducers';
import { EventType } from '../types';

export const getEvents = (state: RootState) => state.events.items;

export const filterEventsByType = (type: EventType) =>
  createSelector([getEvents], (events) =>
    events.filter((event) => event.event_type === type)
  );

export const getAlertsRaised = () =>
  createSelector(
    [filterEventsByType(EventType.ALERT_RAISED)],
    (events) => events
  );

export const getConcernsRaisedBySeverity = () =>
  createSelector([filterEventsByType(EventType.CONCERN_RAISED)], (events) =>
    events.reduce(
      (stat, event) => ({
        ...stat,
        [event.severity]: (stat[event.severity] || 0) + 1,
      }),
      {} as { [key: string]: number }
    )
  );
