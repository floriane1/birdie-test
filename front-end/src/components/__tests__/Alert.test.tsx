import * as React from 'react';
import { shallow, configure } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import Alert from '@App/components/Alert';
import { Event, EventType } from '@App/store//types';

const eventTestA: Event = {
  id: 'event-id-A',
  event_type: EventType.ALERT_RAISED,
  care_recipient_id: 'care-recipient-id-123',
  timestamp: new Date('2019-04-24T02:00:11.187Z'),
};

const eventTestB: Event = {
  id: 'event-id-B',
  event_type: EventType.ALERT_RAISED,
  care_recipient_id: 'care-recipient-id-123',
  timestamp: new Date('2019-05-24T02:00:11.187Z'),
};

const eventTestC: Event = {
  id: 'event-id-C',
  event_type: EventType.CONCERN_RAISED,
  care_recipient_id: 'care-recipient-id-123',
  timestamp: new Date('2019-06-24T02:00:11.187Z'),
  severity: 'high',
};

configure({ adapter: new Adapter() });

describe('Alert', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <Alert
        events={[eventTestA, eventTestB, eventTestC]}
        alerts={[eventTestA, eventTestB]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
