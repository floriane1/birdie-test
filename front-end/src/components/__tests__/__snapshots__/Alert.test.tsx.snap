// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert renders correctly 1`] = `
<styled.div>
  <styled.div>
    <Styled(ExclamationTriangle) />
  </styled.div>
  <div>
    <styled.h1>
      2
       alerts have 
      been raised since 
      June 23, 2019
    </styled.h1>
    <styled.h3>
      The last one was 
      a year ago
    </styled.h3>
  </div>
</styled.div>
`;
